File: src/api/templates.ts
import { apiRequest } from './client';
import type { Template } from '../App';

interface TemplateApiResponse {
  id: string;
  name: string;
  is_default: boolean;
  sections: any[];
}

function toFrontend(t: TemplateApiResponse): Template {
  return {
    id: t.id,
    name: t.name,
    isDefault: t.is_default,
    sections: t.sections,
  };
}

export async function listTemplates(): Promise<Template[]> {
  const data = await apiRequest<TemplateApiResponse[]>('/templates');
  return data.map(toFrontend);
}

export async function createTemplate(
  name: string,
  sections: any[]
): Promise<Template> {
  const data = await apiRequest<TemplateApiResponse>('/templates', {
    method: 'POST',
    body: JSON.stringify({ name, sections }),
  });
  return toFrontend(data);
}

export async function updateTemplate(
  id: string,
  updates: { name?: string; sections?: any[] }
): Promise<Template> {
  const data = await apiRequest<TemplateApiResponse>(`/templates/${id}`, {
    method: 'PUT',
    body: JSON.stringify(updates),
  });
  return toFrontend(data);
}

export async function deleteTemplate(id: string): Promise<void> {
  await apiRequest(`/templates/${id}`, { method: 'DELETE' });
}

File: backend/app/routers/templates.py
import uuid

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.models.template import Template
from app.schemas.template import TemplateCreate, TemplateResponse, TemplateUpdate

router = APIRouter()


@router.get("", response_model=list[TemplateResponse])
async def list_templates(db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(Template))
    templates = result.scalars().all()
    return templates


@router.get("/{template_id}", response_model=TemplateResponse)
async def get_template(template_id: str, db: AsyncSession = Depends(get_db)):
    template = await db.get(Template, template_id)
    if not template:
        raise HTTPException(status_code=404, detail="Template not found")
    return template


@router.post("", response_model=TemplateResponse, status_code=201)
async def create_template(body: TemplateCreate, db: AsyncSession = Depends(get_db)):
    template = Template(
        id=str(uuid.uuid4()),
        name=body.name,
        is_default=False,
        sections=[s.model_dump() for s in body.sections],
    )
    db.add(template)
    await db.flush()
    await db.refresh(template)
    return template


@router.put("/{template_id}", response_model=TemplateResponse)
async def update_template(
    template_id: str,
    body: TemplateUpdate,
    db: AsyncSession = Depends(get_db),
):
    template = await db.get(Template, template_id)
    if not template:
        raise HTTPException(status_code=404, detail="Template not found")

    if body.name is not None:
        template.name = body.name
    if body.sections is not None:
        template.sections = [s.model_dump() for s in body.sections]

    await db.flush()
    await db.refresh(template)
    return template


@router.delete("/{template_id}")
async def delete_template(template_id: str, db: AsyncSession = Depends(get_db)):
    template = await db.get(Template, template_id)
    if not template:
        raise HTTPException(status_code=404, detail="Template not found")
    if template.is_default:
        raise HTTPException(status_code=400, detail="Cannot delete default template")

    await db.delete(template)
    return {"status": "deleted"}
